src/fv/share/tdymeshcustom.c

```diff
@@ -0,0 +1,1315 @@
+  PetscInt nfaces2 = GetNumFacesForCellType(cell_type2);
+
+  for (PetscInt iface2=0; iface2<nfaces2; iface2++) {
+    PetscInt face_id2 = cell_to_face[iface2][cell_id2];
+    PetscInt nvertices2 = GetNumOfVerticesOfIthFacesForCellType(cell_type2, iface2);
+
+    // Check if the number of vertices forming the face are same
+    if (nvertices == nvertices2) {
+
+      // Now check if all the vertices of 'iface' are present in 'face2'
+      PetscInt num_match = 0;
+
+      for (PetscInt ivertex=0; ivertex<nvertices; ivertex++) {
+          PetscInt vertex_id = face_to_vertex[ivertex][face_id];
+          PetscBool vertex_found = PETSC_FALSE;
+
+        for (PetscInt ivertex2=0; ivertex2<nvertices2; ivertex2++) {
+          PetscInt vertex_id2 = face_to_vertex[ivertex2][face_id2];
+          if (vertex_id == vertex_id2) {
+            vertex_found = PETSC_TRUE;
+            num_match++;
+            break;
+          }
+        }
+        if (!vertex_found) {
+          break;
+        }
+      }
+
+      if (num_match == nvertices) {
+        corresponding_face_id = iface2;
+        break;
+      }
+    }
+  }
+
+  PetscFunctionReturn(corresponding_face_id);
+}
+
+/// Create maps between cells, faces, and vertices. So, the duplicate face needs to be removed.
+///
+/// @param [in] ugrid A TDyUGrid struct
+/// @param [out] cell_to_face A map to get face ID for a iface-th of a cell
+/// @param [out] face_to_cell A map to get IDs of upwind and downnwind cells
+/// @param [out] face_to_vertex A map to get face vertex IDs forming a face
+///
+/// @returns 0 on success, or a non-zero error code on failure
+static PetscErrorCode SetupMaps_C2F_F2C_F2V(TDyUGrid *ugrid, PetscInt **cell_to_face, PetscInt **face_to_cell, PetscInt **face_to_vertex) {
+
+  PetscErrorCode ierr;
+  PetscInt ngmax = ugrid->num_cells_global;
+
+  PetscInt *vertex_ids;
+  ierr = TDyAllocate_IntegerArray_1D(&vertex_ids, 4); CHKERRQ(ierr);
+
+  PetscInt face_count=0;
+  for (PetscInt icell=0; icell<ngmax; icell++) {
+    PetscInt num_vertices = ugrid->cell_num_vertices[icell];
+    TDyCellType cell_type = GetCellType(num_vertices);
+    PetscInt nfaces = GetNumFacesForCellType(cell_type);
+
+    for (PetscInt iface=0; iface<nfaces; iface++) {
+      cell_to_face[iface][icell] = face_count;
+      face_to_cell[0][face_count] = icell;
+
+      PetscInt nvertices = GetNumOfVerticesOfIthFacesForCellType(cell_type, iface);
+      ierr = GetFaceVertices(cell_type, iface, vertex_ids); CHKERRQ(ierr);
+
+      for (PetscInt ivertex=0; ivertex<nvertices; ivertex++) {
+
+        PetscInt vertex_id_relative = vertex_ids[ivertex];
+        PetscInt vertex_id = ugrid->cell_vertices[icell][vertex_id_relative];
+
+        face_to_vertex[ivertex][face_count] = vertex_id;
+
+        if (vertex_id > -1) {
+          ugrid->face_to_vertex_natural[ivertex][face_count] = ugrid->vertex_ids_natural[vertex_id];
+        }
+      }
+
+      face_count++;
+    }
+  }
+  free(vertex_ids);
```

Comment:
This is another example of how we need to unify our memory allocation code (it's not clear in this context that `TDyAllocate_IntegerArray_1D` above uses `malloc` to allocate the memory being freed here).